pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -470,262 +559,611 @@ fn storage_sync_loop<
+                            sync_id.tenant_id, sync_id.timeline_id, e
+                        );
+                        None
+                    }
+                };
+                (sync_id, state_update)
+            }
         })
         .collect::<FuturesUnordered<_>>();
 
     let mut new_timeline_states: HashMap<
         ZTenantId,
         HashMap<ZTimelineId, TimelineSyncStatusUpdate>,
     > = HashMap::with_capacity(max_concurrent_sync);
-    while let Some((sync_id, state_update)) = task_batch.next().await {
+    while let Some((sync_id, state_update)) = sync_results.next().await {
         debug!("Finished storage sync task for sync id {}", sync_id);
         if let Some(state_update) = state_update {
-            let ZTenantTimelineId {
-                tenant_id,
-                timeline_id,
-            } = sync_id;
             new_timeline_states
-                .entry(tenant_id)
+                .entry(sync_id.tenant_id)
                 .or_default()
-                .insert(timeline_id, state_update);
+                .insert(sync_id.timeline_id, state_update);
         }
     }
 
-    LoopStep::SyncStatusUpdates(new_timeline_states)
+    ControlFlow::Continue(new_timeline_states)
 }
 
-async fn process_task<
-    P: std::fmt::Debug + Send + Sync + 'static,
-    S: RemoteStorage<StoragePath = P> + Send + Sync + 'static,
->(
+async fn process_sync_task<P, S>(
     conf: &'static PageServerConf,
-    remote_assets: Arc<(S, RemoteIndex)>,
-    task: SyncTask,
+    storage: Arc<S>,
+    index: RemoteIndex,
     max_sync_errors: NonZeroU32,
-) -> Option<TimelineSyncStatusUpdate> {
-    if task.retries > max_sync_errors.get() {
-        error!(
-            "Evicting task {:?} that failed {} times, exceeding the error threshold",
-            task.kind, task.retries
-        );
-        FATAL_TASK_FAILURES.inc();
-        // FIXME (rodionov) this can potentially leave holes in timeline uploads
-        //    planneed to be fixed as part of https://github.com/zenithdb/zenith/issues/977
-        return None;
-    }
+    sync_id: ZTenantTimelineId,
+    task: SyncTask,
+) -> Option<TimelineSyncStatusUpdate>
+where
+    P: Debug + Send + Sync + 'static,
+    S: RemoteStorage<StoragePath = P> + Send + Sync + 'static,
+{
+    let sync_start = Instant::now();
+    let current_remote_timeline = { index.read().await.timeline_entry(&sync_id).cloned() };
+
+    let task = match validate_task_retries(sync_id, task, max_sync_errors) {
+        ControlFlow::Continue(task) => task,
+        ControlFlow::Break(aborted_task) => {
+            match aborted_task {
+                SyncTask::Download(_) => {
+                    index
+                        .write()
+                        .await
+                        .set_awaits_download(&sync_id, false)
+                        .ok();
+                }
+                SyncTask::Upload(failed_upload_data) => {
+                    if let Err(e) = update_remote_data(
```

Comment:
We update the remote index first before updating anything on S3, so that bit makes sense even without working S3.

I would also like to leave it, since as we've noticed, some S3 quirks can lead to situations where some files are uploaded just fine, while the others aren't.
This way, we at least have the chance to know that something went wrong and we have "holes" in the timeline.

Reply:
IMO this approach is rather temporary but will work for now. I'm leaning towards infinite retries without additional data on the remote storage. Because we can transfer the timeline to different pageserver and as I can see there is nothing it can do about such missing files so they'll reside in metadata forever.

Reply:
Yes, the eviction is very raw and same is the information about the "holes" now: it's just present but we don't do any actions on that besides adding more layers to it. 
We have a separate task for dealing with eviction cases, so should not be an issue right now.